Source/Details/_ASCollectionReusableView.h

```diff
@@ -18,15 +18,41 @@
 #import <UIKit/UIKit.h>
 #import <AsyncDisplayKit/ASBaseDefines.h>
 
-@class ASCellNode, ASCollectionElement;
+@class ASCollectionElement;
+
+/**
+ * Attempts to cast x to _ASCollectionReusableView, or nil if not possible.
+ *
+ * Since _ASCollectionReusableView is not available for subclassing (see below),
+ * comparing x's and _ASCollectionReusableView's classes is faster than calling -isKindOfClass: on x.
+ */
+#define ASCollectionReusableViewCast(x) ({ \
+  id __var = x; \
+  ((_ASCollectionReusableView *) (x.class == [_ASCollectionReusableView class] ? __var : nil)); \
+})
```

Comment:
I see that @appleguy started this pattern at line 1100, and that you've formalized it, but it strikes me as serious over-optimization, and I think we should simply remove it and use whatever is most concise at the call site. `-isKindOfClass:` is extremely fast. Here's the source code, from https://opensource.apple.com/source/objc4/objc4-709/runtime/NSObject.mm

```
- (BOOL)isKindOfClass:(Class)cls {
    for (Class tcls = [self class]; tcls; tcls = tcls->superclass) {
        if (tcls == cls) return YES;
    }
    return NO;
}
```

It only includes one method call – which is unavoidable – and otherwise just uses raw pointer comparison memory reads.